File: 백준/Gold/10830. 행렬 제곱/행렬 제곱.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {
	private static int N;
	private static int[][] matrix;

	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringBuilder sb = new StringBuilder();
		StringTokenizer st = new StringTokenizer(br.readLine());

		N = Integer.parseInt(st.nextToken());
		long B = Long.parseLong(st.nextToken());
		matrix = new int[N][N];
		for (int i = 0; i < N; i++) {
			st = new StringTokenizer(br.readLine());

			for (int j = 0; j < N; j++) {
				matrix[i][j] = Integer.parseInt(st.nextToken());
			}
		}

		int[][] res = pow(B);
		for (int i = 0; i < N; i++) {
			for (int j = 0; j < N; j++) {
				sb.append(res[i][j] % 1000).append(' ');
			}
			sb.append('\n');
		}

		System.out.println(sb);
	}

	private static int[][] pow(long cnt) {
		if (cnt == 1) {
			return matrix;
		}

		int[][] A = pow(cnt / 2);

		if (cnt % 2 == 0) {
			return cal(A, A);
		} else {
			return cal(A, cal(A, matrix));
		}
	}

	private static int[][] cal(int[][] A, int[][] B) {
		int[][] arr = new int[N][N];

		for (int i = 0; i < N; i++) {
			for (int j = 0; j < N; j++) {
				for (int k = 0; k < N; k++) {
					arr[i][j] += (A[i][k] * B[k][j]);
					arr[i][j] %= 1000;
				}
			}
		}

		return arr;
	}
}

File: 백준/Gold/10830. 행렬 제곱/행렬 제곱.kt
import java.io.BufferedReader
import java.io.InputStreamReader
import java.lang.StringBuilder
import java.util.LinkedList
import java.util.StringTokenizer

var N = 0
lateinit var matrix: Array<IntArray>

fun main() = with(BufferedReader(InputStreamReader(System.`in`))) {
    val sb = StringBuilder()
    var st = StringTokenizer(readLine())

    N = st.nextToken().toInt()
    val B = st.nextToken().toLong()
    matrix = Array(N) { IntArray(N) }
    for (i in 0 until N) {
        st = StringTokenizer(readLine())

        for (j in 0 until N) {
            matrix[i][j] = st.nextToken().toInt()
        }
    }

    val res = pow(B)
    for (i in 0 until N) {
        for (j in 0 until N) {
           sb.append(res[i][j] % 1000).append(' ')
        }
        sb.append('\n');
    }

    println(sb)
}

fun pow(cnt: Long): Array<IntArray> {
    if (cnt == 1L)
        return matrix

    val A = pow(cnt / 2)

    if (cnt % 2 == 0L)
        return cal(A, A)
    else
        return cal(A, cal(A, matrix))
}

fun cal(A: Array<IntArray>, B: Array<IntArray>): Array<IntArray> {
    val arr = Array(N) { IntArray(N) }
    for (i in 0 until N) {
        for (j in 0 until N) {
            for (k in 0 until N) {
                arr[i][j] += A[i][k] * B[k][j];
                arr[i][j] %= 1000;
            }
        }
    }

    return arr;
}